[TreeView] Add label editing feature (#13388)
mui/mui-x: docs/data/tree-view/rich-tree-view/editing/editing.md

---
productId: x-tree-view
title: Rich Tree View - Focus
githubLabel: 'component: tree view'
packageName: '@mui/x-tree-view'
waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
packageName: '@mui/x-tree-view'
---

# Rich Tree View - label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

:::warning
this doc page is stip WIP
:::

To edit a label double click the item or press Enter on the keyboard

Enter again after editing will save the new label. Bluring or pressing escape will cancel the editing

> {{"demo": "LabelEditing.js"}}

Reviewer: A demo showing how to only edit leaves would be an interesting addition (you can check the drag & drop PR which has a very similar demo)

---
productId: x-tree-view
title: Rich Tree View - Focus
githubLabel: 'component: tree view'
packageName: '@mui/x-tree-view'
waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
packageName: '@mui/x-tree-view'
---

# Rich Tree View - label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

:::warning
this doc page is stip WIP
:::

## Enable editing on certain items

You can use the `isItemEditable` prop to enable editing on certain items.

If an item is editable, the editing state can be toggled by double clicking on it, or by pressing <kbd class="key">Enter</kbd> on the keyboard when the item is in focus.

Once an item is in editing state, the value of the label can be edited. Pressing <kbd class="key">Enter</kbd> again or bluring the item will save the new value. Pressing <kbd class="key">Esc</kbd> will cancel the action and restore the item to its original state.

{{"demo": "LabelEditing.js"}}

## Change the default behavior

The default behavior on bluring the tree item that is being edited is to save the new value if there is one. This can be customized using the `slotProps` of the `TreeItem2`

{{"demo": "CustomBehavior.js"}}

> ## Limit editing to leaves

Reviewer: I would make this section a child of `## Enable editing on certain items`

Author: > We could consider dropping the interactions in favor of only using publicAPI methods at some point

I think that makes sense. 

> I'm not entirely sure how the DX should look like at this level...

We could have a [similar DX](https://mui.com/x/react-data-grid/api-object/#inside-the-data-grid) to the grid, but it's slightly confusing IMO :(

Reviewer: We already return `publicAPI` from `useTreeItem2` :+1:

---
productId: x-tree-view
title: Rich Tree View - Editing
githubLabel: 'component: tree view'
packageName: '@mui/x-tree-view'
waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
packageName: '@mui/x-tree-view'
---

# Rich Tree View - label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

> :::warning

Reviewer: I think you can drop this comment now that we are close from the open review
Tbh I would even open for review today to gather feedbacks

this doc page is stip WIP
:::

> ## Enable editing on certain items

Reviewer: ```suggestion
## Enable editing
```

I think we should either

1. Add a prop `labelEditing?: boolean` and if `isItemEditable` is not defined then all items are editable
2. Keep the current DX, in which case we rename this doc section "Enable label editing" 

Because right now I feel like we miss the most basic use case: how to enable label editing for every item :laughing:

Author: I renamed this section for now, but if we feel like we want to introduce another prop, I'm fine with that too 👌

You can use the `isItemEditable` prop to enable editing on certain items.

If an item is editable, the editing state can be toggled by double clicking on it, or by pressing <kbd class="key">Enter</kbd> on the keyboard when the item is in focus.

Once an item is in editing state, the value of the label can be edited. Pressing <kbd class="key">Enter</kbd> again or bluring the item will save the new value. Pressing <kbd class="key">Esc</kbd> will cancel the action and restore the item to its original state.

{{"demo": "LabelEditing.js"}}

### Limit editing to leaves

You can limit the editing to just the leaves of the tree.

{{"demo": "EditLeaves.js"}}

## Track item label change

Use the `onItemLabelChange` prop to trigger an action when the label of an item changes.

{{"demo": "EditingCallback.js"}}

## Change the default behavior

The default behavior on bluring the tree item that is being edited is to save the new value if there is one. This can be customized using the `slotProps` of the `TreeItem2`

{{"demo": "CustomBehavior.js"}}

## Enable editing using only icons

> The demo below shows how to entirely override the editing behavior, and implementit using icons.

Reviewer: ```suggestion
The demo below shows how to entirely override the editing behavior, and implement it using icons.
```

{{"demo": "EditWithIcons.js"}}

## Create a custom labelInput

The demo below shows how to use a different component in the `labelInput` slot.

{{"demo": "CustomLabelInput.js"}}

## Imperative API

:::success
To use the `apiRef` object, you need to initialize it using the `useTreeViewApiRef` hook as follows:

```tsx
const apiRef = useTreeViewApiRef();

return <RichTreeView apiRef={apiRef} items={ITEMS}>;
```

When your component first renders, `apiRef` will be `undefined`.
After this initial render, `apiRef` holds methods to interact imperatively with the Tree View.
:::

### Change the label of an item

Use the `setItemExpansion` API method to change the expansion of an item.

```ts
apiRef.current.updateItemLabel(

>   // The id of the item to expand or collapse

Reviewer: The description needs to be updated

packageName: '@mui/x-tree-view'
---

# Rich Tree View - label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

## Enable label editing

You can use the `isItemEditable` prop to enable editing on certain items.

:::success
You can enable editing on all items by always returning true

```jsx
<RichTreeView items={ITEMS} isItemEditable={() => true} />
```

:::

If an item is editable, the editing state can be toggled by double clicking on it, or by pressing <kbd class="key">Enter</kbd> on the keyboard when the item is in focus.

Once an item is in editing state, the value of the label can be edited. Pressing <kbd class="key">Enter</kbd> again or bluring the item will save the new value. Pressing <kbd class="key">Esc</kbd> will cancel the action and restore the item to its original state.

{{"demo": "LabelEditing.js"}}

### Limit editing to leaves

You can limit the editing to just the leaves of the tree.

{{"demo": "EditLeaves.js"}}

## Track item label change

Use the `onItemLabelChange` prop to trigger an action when the label of an item changes.

{{"demo": "EditingCallback.js"}}

## Change the default behavior

> The default behavior on bluring the tree item that is being edited is to save the new value if there is one. This can be customized using the `slotProps` of the `TreeItem2`

Reviewer 2: ```suggestion
By default, blurring the tree item saves the new value if there is one.
To modify this behavior, use the `slotProps` of the `TreeItem2`.
```

---
productId: x-tree-view
title: Rich Tree View - Editing
githubLabel: 'component: tree view'
packageName: '@mui/x-tree-view'
waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
packageName: '@mui/x-tree-view'
---

# Rich Tree View - label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

## Enable label editing

You can use the `isItemEditable` prop to enable editing on certain items.

:::success
You can enable editing on all items by always returning true

```jsx

> <RichTreeView items={ITEMS} isItemEditable={() => true} />

Author: Makes sense 🤔 I'm wondering about the naming though in this case

> <RichTreeView items={ITEMS} isItemEditable={() => true} />

Reviewer 3: Do you think the current prop name wouldn't be clear enough for both a callback and boolean versions? 🤔

> <RichTreeView items={ITEMS} isItemEditable={() => true} />

Author: My only worry is that the naming is singular and technically if you pass `true` it's valid for **all** the items. But I don't think I can come up with a better name, so I'm okay with leaving it as it is if you don't think it's a problem 👍

> <RichTreeView items={ITEMS} isItemEditable={() => true} />

Reviewer 3: Nor can I think of anything better. 🙈 🤷

```

:::

If an item is editable, the editing state can be toggled by double clicking on it, or by pressing <kbd class="key">Enter</kbd> on the keyboard when the item is in focus.

Once an item is in editing state, the value of the label can be edited. Pressing <kbd class="key">Enter</kbd> again or bluring the item will save the new value. Pressing <kbd class="key">Esc</kbd> will cancel the action and restore the item to its original state.

{{"demo": "LabelEditing.js"}}

### Limit editing to leaves

You can limit the editing to just the leaves of the tree.

{{"demo": "EditLeaves.js"}}

## Track item label change

Use the `onItemLabelChange` prop to trigger an action when the label of an item changes.

{{"demo": "EditingCallback.js"}}

## Change the default behavior

> By default, blurring the tree item saves the new value if there is one.

Author: There is no equality check, but we don't save is there is no value. A rephrasing would probably be nice here 🙈 
How about:
```suggestion
By default, blurring the tree item saves the new value if it's not empty.
```

---
productId: x-tree-view
title: Rich Tree View - Editing
githubLabel: 'component: tree view'
packageName: '@mui/x-tree-view'
waiAria: https://www.w3.org/WAI/ARIA/apg/patterns/treeview/
packageName: '@mui/x-tree-view'
---

# Rich Tree View - Label editing

<p class="description">Learn how to edit the label of Tree View items.</p>

## Enable label editing

> You can use the `isItemEditable` prop to enable editing on certain items.

Reviewer: I would remove the `:::success` and just show a 1st demo with all items enabled and then have a section `# Limit editing to some items` or something like that (or in the same section if you prefer)

Having the most basic example in a `:::success` is not optimal IMHO.

Something like that:

```
## Enable label editing

You can use the `isItemEditable` prop to enable editing.
If set to `true`, this prop will enable label editing on all items:

{{"demo": "LabelEditingAllItems.js"}}

If you pass a method to `isItemEditable`, only the items for which the method returns `true` will be editable:

{{"demo": "LabelEditingSomeItems.js"}}

:::success
If an item is editable, the editing state can ........
:::
```

I moved the UX description in a `:::success` because for me it is out of the main goal of the section (which is how to enable the feature).
If you prefer you can create a new section, but I don't think it's worth it

Author: Done! I made some slight changes to your suggestion though:
- `Enable editing` is its own section, and the success callout below the first demo describes the UX/interactions
- `Limit editing to some items` is the second section, and `Limit editing to leaves` is an h3 within that section

WDYT?

Reviewer: Looks good to me :+1:
